Refactor to use Mdbook and introduction

baszalmstra opened the pull request.

baszalmstra committed:
mdbook implementation of the book

baszalmstra committed:
merged chapters

baszalmstra committed:
Updated introduction

Wodann commented on book.toml:
```diff
@@ -0,0 +1,6 @@
+[book]
+authors = ["Bas Zalmstra"]
```
```suggestion
authors = ["Bas Zalmstra", "Remco Kuijper"]
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+*Mun* is an embedable scripting language designed for developer productivity. 
+
+* **Ahead of time compiled**
```
```suggestion
* **Ahead of time compilation**
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+*Mun* is an embedable scripting language designed for developer productivity. 
+
+* **Ahead of time compiled**
+  Mun is ahead of time (AOT) compiled instead of interpreted or JITed. This
```
```suggestion
  Mun is compiled ahead of time (AOT), as opposed to being interpreted or compiled just in time (JIT).
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+* **Ahead of time compiled**
+  Mun is ahead of time (AOT) compiled instead of interpreted or JITed. This
+  enables catching a lot of errors at compile time instead of at runtime. AOT
```
```suggestion
  By detecting type errors during AOT compilation, an entire class of runtime errors is eliminated.
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+* **Ahead of time compiled**
+  Mun is ahead of time (AOT) compiled instead of interpreted or JITed. This
+  enables catching a lot of errors at compile time instead of at runtime. AOT
+  compilation allows developers to stay in their IDE instead of having to debug
```
```suggestion
  Whereas developers of interpreted or JIT compiled languages have to constantly switch between their IDE and target application to debug runtime errors, an AOT compiled language allows developers to stay within the comfort of their IDE.
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+*Mun* is an embedable scripting language designed for developer productivity. 
+
+* **Ahead of time compiled**
```
Maybe add something along the lines of:

```
AOT compilation results in machine code that can be natively executed on a platform,
guaranteeing the best possible performance. Interpreted languages suffer a runtime
penalty as every bytecode instruction needs to be translated to an equivalent machine
code instruction. JIT compiled languages try to overcome this performance deficiency
by optimising their program during an extra runtime compilation.
```

to show that AOT has performance benefits.

Wodann replied:
We could also add that some platforms don't allow for JIT compilation as they prevent write access to executable memory.

baszalmstra replied:
I think that should be covered in the Performance section in that case.

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  the target application to check for runtime errors. 
+
+* **Statically typed**
+  Mun uses types that are resolved at compilation time instead of at runtime.
```
```suggestion
  Mun resolves types at compilation time instead of at runtime, resulting in
  immediate feedback when writing code and opening the door for powerful
  refactoring tools.
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+* **Statically typed**
+  Mun uses types that are resolved at compilation time instead of at runtime.
+  This enables powerful refactoring tools and better error detection at the time
```
```suggestion
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+* **Statically typed**
+  Mun uses types that are resolved at compilation time instead of at runtime.
+  This enables powerful refactoring tools and better error detection at the time
+  of writing the code vs running the code.
```
```suggestion
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  of writing the code vs running the code.
+
+* **First class hot-reloading**
+  Mun is written from the ground up with hot-reloading in mind. Hot-reloading
```
```suggestion
  Every aspect of Mun is designed with hot reloading in mind. Hot reloading is the
  process of changing code and resources of a live application, removing the need
  to start, stop and recompile an application whenever a function or value needs to
  be changed.```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+* **First class hot-reloading**
+  Mun is written from the ground up with hot-reloading in mind. Hot-reloading
+  enables developers to alter the code while its executing. There no longer is a
```
```suggestion
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+* **First class hot-reloading**
+  Mun is written from the ground up with hot-reloading in mind. Hot-reloading
+  enables developers to alter the code while its executing. There no longer is a
+  need to start/stop the application, changing functions, values or classes take
```
```suggestion
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  Mun is written from the ground up with hot-reloading in mind. Hot-reloading
+  enables developers to alter the code while its executing. There no longer is a
+  need to start/stop the application, changing functions, values or classes take
+  effect instantly without restarting the target application. 
```
```suggestion
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  effect instantly without restarting the target application. 
+
+* **Performance** 
+  Mun compiles the code to optimized machine code using LLVM. Together with
```
```suggestion
  Mun compiles code to optimized machine code using LLVM. In combination with
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+* **Performance** 
+  Mun compiles the code to optimized machine code using LLVM. Together with
+  static typing this enables native performance. Hot-reloading is enabled by
```
```suggestion
  static typing this guarantees native runtime performance. Hot reloading does
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+* **Performance** 
+  Mun compiles the code to optimized machine code using LLVM. Together with
+  static typing this enables native performance. Hot-reloading is enabled by
+  default which adds a little bit of overhead at runtime, however, for
```
```suggestion
  introduce a slight runtime overhead, but it can be disabled for production
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  Mun compiles the code to optimized machine code using LLVM. Together with
+  static typing this enables native performance. Hot-reloading is enabled by
+  default which adds a little bit of overhead at runtime, however, for
+  production builds hot-reloading can also be turned off which removes all
```
```suggestion
  builds to ensure the best possible runtime performance.
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  static typing this enables native performance. Hot-reloading is enabled by
+  default which adds a little bit of overhead at runtime, however, for
+  production builds hot-reloading can also be turned off which removes all
+  overhead.
```
```suggestion
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+  supported compiler platform. 
+
+* **Powerful IDE integration**
+  The Mun language and compiler framework are designed from the ground up to
```
```suggestion
  The Mun language and compiler framework are designed to
```

Wodann replied:
From the ground up is a separate statement, and as such its meaning doesn't work as a lead up: https://idioms.thefreedictionary.com/from+the+ground+up

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,39 @@
+
+* **Powerful IDE integration**
+  The Mun language and compiler framework are designed from the ground up to
+  support queries on the source code which enable powerful IDE integration like
```
```suggestion
  support source code queries, allowing for powerful IDE integrations such as
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+# Case Studies
+
+Here are some of the case studies that inform the design choices made in Mun.
```
```suggestion
A collection of case studies that inspired the design choices made in Mun.
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+patterns recur a lot and as such have become standard practice. Introducing
+these practices assists us in daily development a lot, but requires more code
+and complexity than desirable. Having syntactic sugar would greatly help
+reduce complexity in our code base but would also introduce *magic* or custom
```
```suggestion
reduce complexity in our code base, but would also introduce *magic* or custom
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+these practices assists us in daily development a lot, but requires more code
+and complexity than desirable. Having syntactic sugar would greatly help
+reduce complexity in our code base but would also introduce *magic* or custom
+Abbey Games keywords which are foreign to both new developers and IDE's. 
```
```suggestion
keywords that are foreign to both new developers and IDE's. 
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+## Abbey Games
+
+Abbey Games uses Lua as its main gameplay programming language because of Lua's
+live reload the code. This enables very rapid iteration of game code enabling
```
```suggestion
ability to hot reload code. This allows for very rapid iteration of game code, enabling
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+
+Abbey Games uses Lua as its main gameplay programming language because of Lua's
+live reload the code. This enables very rapid iteration of game code enabling
+game programmers & designers alike to quickly test and tweak systems and
```
```suggestion
gameplay programmers and designers to quickly test and tweak systems and
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+Abbey Games uses Lua as its main gameplay programming language because of Lua's
+live reload the code. This enables very rapid iteration of game code enabling
+game programmers & designers alike to quickly test and tweak systems and
+content. Lua however, is a dynamically typed, Just-in-time compiled language,
```
```suggestion
content. Lua is a dynamically typed, JIT compiled language. Although this has
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+live reload the code. This enables very rapid iteration of game code enabling
+game programmers & designers alike to quickly test and tweak systems and
+content. Lua however, is a dynamically typed, Just-in-time compiled language,
+although this has some definite advantages it also introduces a lot of problems
```
```suggestion
some definite advantages, it also introduces a lot of problems with bigger codebases.
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+game programmers & designers alike to quickly test and tweak systems and
+content. Lua however, is a dynamically typed, Just-in-time compiled language,
+although this has some definite advantages it also introduces a lot of problems
+with a bigger codebase.
```
```suggestion
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+# Types
+
+Mun is a statically typed languages which helps to detect type-related errors in
```
```suggestion
Mun is a statically typed language, which helps to detect type-related errors at
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+# Types
+
+Mun is a statically typed languages which helps to detect type-related errors in
+a program at compile-time. A type error is an invalid operation on a given type
```
```suggestion
compile-time. A type error is an invalid operation on a given type, such as an
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+
+Mun is a statically typed languages which helps to detect type-related errors in
+a program at compile-time. A type error is an invalid operation on a given type
+such as a division by a string, trying to access a field that doesn't exist or
```
```suggestion
integer divided by a string, trying to access a field that doesn't exist, or
```

Wodann commented on src/ch01-00-types.md:
````diff
@@ -0,0 +1,37 @@
+with type information:
+
+```c++
+int foo = 3+4;
````
```suggestion
int foo = 3 + 4;
```

Wodann commented on src/ch01-00-types.md:
````diff
@@ -0,0 +1,37 @@
+int foo = 3+4;
+```
+
+Often the types of variables can be inferred by their usage just as in the
````
```suggestion
However, often variable types can be inferred by their usage, as illustrated
```

Wodann commented on src/ch01-00-types.md:
````diff
@@ -0,0 +1,37 @@
+```
+
+Often the types of variables can be inferred by their usage just as in the
+example above. Mun uses type inferencing to determine the type of a variable at
````
```suggestion
above. Mun uses type inferencing to determine variable types at compile time.
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+
+Often the types of variables can be inferred by their usage just as in the
+example above. Mun uses type inferencing to determine the type of a variable at
+compile time. However, it does force you to explicitly annotate variables in a
```
```suggestion
You are still forced to explicitly annotate variables in a few locations to ensure
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+Often the types of variables can be inferred by their usage just as in the
+example above. Mun uses type inferencing to determine the type of a variable at
+compile time. However, it does force you to explicitly annotate variables in a
+few locations to ensure a contract between interdependent code.
```
```suggestion
a contract between interdependent code.
```

Wodann commented on src/ch01-00-types.md:
````diff
@@ -0,0 +1,37 @@
+
+```mun
+function bar(a:int) {
+    let foo = 3+a
````
```suggestion
    let foo = 3 + a
```

Wodann commented on src/ch01-00-types.md:
````diff
@@ -0,0 +1,37 @@
+```
+
+Here, the parameter `a` must be annotated because it solidifies the signature of
+the function. However, the type of `foo` can be inferred through it's usage.
````
```suggestion
the function. The type of `foo` can be inferred through its usage.
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+double-precision IEEE 64-bit floating point number and an `int` represents a
+64-bit integral number.
+
+In Mun an `int` can be explicitly casted to `float` but the reverse is not true:
```
```suggestion
In Mun an `int` can be explicitly cast to `float`, but the reverse is not true.
```

Wodann replied:
Do you mean `implicitly cast`? Otherwise the next line should be changed to read `and can therefore not be cast explicitly`.

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+64-bit integral number.
+
+In Mun an `int` can be explicitly casted to `float` but the reverse is not true:
+assigning a `float` to an `int` might cause loss of precision and is therefor
```
```suggestion
Assigning a `float` to an `int` might cause loss of precision and can therefore
```

Wodann commented on src/ch01-00-types.md:
```diff
@@ -0,0 +1,37 @@
+
+In Mun an `int` can be explicitly casted to `float` but the reverse is not true:
+assigning a `float` to an `int` might cause loss of precision and is therefor
+not allowed implicitly.
```
```suggestion
not be cast implicitly.
```

Wodann replied:
`implicit` or `explicit`?

Wodann commented on src/ch02-00-types.md:
```diff
@@ -0,0 +1 @@
+# Types
```
Is this file not a copy of chapter 1?

baszalmstra replied:
Yes, it should not be here.

baszalmstra committed:
Update book.toml

Co-Authored-By: Wodann <Wodann@users.noreply.github.com>

baszalmstra committed:
Apply suggestions from code review

Co-Authored-By: Wodann <Wodann@users.noreply.github.com>

baszalmstra committed:
properly incorporated Remcos changes

baszalmstra commented:
I incorporated most of your changes @Wodann , except for "We could also add that some platforms don't allow for JIT compilation as they prevent write access to executable memory." This would answer the question why not use JIT'ing at runtime. AOT could still be combined with JIT (see typescript for example). However, I don't think we have to answer this question in the introduction so I choose to omit it.

I also moved some stuff from AOT to Performance. 

An new look at the changes would be appreciated! :)

baszalmstra committed:
rewrapped some lines to ensure lines are no longer than 80 chars

baszalmstra committed:
Merge remote-tracking branch 'upstream/master' into mdbook

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,40 @@
+> Note: Mun & this book are currently under active development, any and all
+> content of this book not final and can still change.
+
+*Mun* is an embedable scripting language designed for developer productivity. 
```
```suggestion
*Mun* is an embeddable scripting language designed for developer productivity. 
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,40 @@
+  Every aspect of Mun is designed with hot reloading in mind. Hot reloading is
+  the process of changing code and resources of a live application, removing the
+  need to start, stop and recompile an application whenever a function or value
+  needs to be changed.
```
```suggestion
  is changed.
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,40 @@
+  needs to be changed.
+
+* **Performance**  
+  Mun compiles code to optimized machine code using LLVM. Mun's AOT compilation
```
```suggestion
  AOT compilation combined with static typing ensure that Mun is compiled to machine code
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,40 @@
+
+* **Performance**  
+  Mun compiles code to optimized machine code using LLVM. Mun's AOT compilation
+  and static typing result in machine code that can be natively executed on a
```
```suggestion
  that can be natively executed on any target platform. LLVM is used for compilation
```

Wodann commented on src/ch00-00-introduction.md:
```diff
@@ -0,0 +1,40 @@
+* **Performance**  
+  Mun compiles code to optimized machine code using LLVM. Mun's AOT compilation
+  and static typing result in machine code that can be natively executed on a
+  platform, guaranteeing the best possible performance. Hot reloading does
```
```suggestion
  and optimization, guaranteeing the best possible performance. Hot reloading does
```

Wodann commented on src/ch00-01-case-studio-abbey-games.md:
```diff
@@ -0,0 +1,37 @@
+## Abbey Games
+
+Abbey Games uses Lua as its main gameplay programming language because of Lua's
+ability to hot reload code. This allows for very rapid iteration of game code,
```
```suggestion
ability to hot reload code. This allows for rapid iteration of game code,
```

baszalmstra committed:
Apply suggestions from code review

Co-Authored-By: Wodann <Wodann@users.noreply.github.com>
